hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpc/nio/NioAsyncSocket.java

```diff
@@ -49,12 +52,37 @@
 import static java.nio.channels.SelectionKey.OP_WRITE;
 
 /**
- * Nio version of the {@link AsyncSocket}.
+ * Nio implementation of the {@link AsyncSocket}.
  */
 @SuppressWarnings({"checkstyle:DeclarationOrder", "checkstyle:VisibilityOrder"})
 public final class NioAsyncSocket extends AsyncSocket {
+
     private static final int DEFAULT_UNFLUSHED_BUFS_CAPACITY = 2 << 16;
 
+    private static final SocketOption<Integer> TCP_KEEPCOUNT = findSocketOption("TCP_KEEPCOUNT");
+    private static final SocketOption<Integer> TCP_KEEPIDLE = findSocketOption("TCP_KEEPIDLE");
+    private static final SocketOption<Integer> TCP_KEEPINTERVAL = findSocketOption("TCP_KEEPINTERVAL");
+    private static final AtomicBoolean TCP_KEEPCOUNT_PRINTED = new AtomicBoolean();
+    private static final AtomicBoolean TCP_KEEPIDLE_PRINTED = new AtomicBoolean();
+    private static final AtomicBoolean TCP_KEEPINTERVAL_PRINTED = new AtomicBoolean();
+
+    private static SocketOption<Integer> findSocketOption(String fieldName) {
+        Class extendedSocketOptions = ExtendedSocketOptions.class;
+        try {
+            Field field = extendedSocketOptions.getDeclaredField(fieldName);
+            return (SocketOption) field.get(null);
+        } catch (Exception e) {
+            return null;
+        }
+    }
```

Comment:
Maybe we can check java version instead of this? If java version is less than 11 we can print those warnings?

Reply:
If we want to go with this, I would extract this method to a util method so that `NioAsyncServerSocket` also uses that, instead of duplicating this logic in a static block.

Reply:
I don't see why that should be better. Just check if the field exists.

Reply:
Duplication is underrated :) I moved it to a utility class, but it is false to assume that duplication is always bad. Often the solution to remove duplication is worse than the duplication itself.